src/lossy_compression/value_representation/serf_qt.zig

```diff
@@ -0,0 +1,325 @@
+// limitations under the License.
+
+//! Implementation of the SERF-QT algorithm from the paper
+//! "Li, Ruiyuan, Zechao Chen, Ruyun Lu, Xiaolong Xu, Guangchao Yang, Chao Chen, Jie Bao, and Yu Zheng.
+//! Serf: Streaming Error-Bounded Floating-Point Compression.
+//! ACM SIGMOD 2025.
+//! https://doi.org/10.1145/3725353.
+
+const std = @import("std");
+const math = std.math;
+const mem = std.mem;
+const io = std.io;
+const ArrayList = std.ArrayList;
+const testing = std.testing;
+const Allocator = std.mem.Allocator;
+
+const tersets = @import("../../tersets.zig");
+const shared_functions = @import("../../utilities/shared_functions.zig");
+const configuration = @import("../../configuration.zig");
+const tester = @import("../../tester.zig");
+
+const Error = tersets.Error;
+const Method = tersets.Method;
+
+/// Compress `uncompressed_values` within error_bound using "Serf-QT". The function writes the
+/// result to `compressed_values`. The `compressed_values` includes an `EliasGamma` encoding of
+/// the quantized elements. The `allocator` is used for memory management of intermediates containers
+/// and the `method_configuration` parser. The function expects an `AbsoluteErrorBound` configuration.
+/// If an error occurs it is returned.
+pub fn compress(
+    allocator: Allocator,
+    uncompressed_values: []const f64,
+    compressed_values: *ArrayList(u8),
+    method_configuration: []const u8,
+) Error!void {
+    const parsed_configuration = try configuration.parse(
+        allocator,
+        configuration.AbsoluteErrorBound,
+        method_configuration,
+    );
+
+    const error_bound: f32 = parsed_configuration.abs_error_bound;
+
+    // Ensure the compressed values are not empty.
+    if (uncompressed_values.len == 0) return Error.UnsupportedInput;
+
+    // Avoid edge cases by defining bucket size slightly larger than 2 * error_bound.
+    const bucket_size: f32 = 1.99 * error_bound;
+    // Append the minimum value to the header of the compressed values.
+    try shared_functions.appendValue(f32, bucket_size, compressed_values);
+
+    //Intermediate quantized values.
+    var quantized_values = ArrayList(u64).init(allocator);
+    defer quantized_values.deinit();
+
+    // Assume the first value is zero for simplicity (Section 4.1).
+    var previous_value: f64 = 0.0;
+
+    // Quantize each value by mapping it to a discrete bucket index.
+    // If the error_bound is zero, we compute the difference between the
+    // value and the minimum value, ensuring all resulting integers are >= 0.
+    // For non-zero error_bound, we apply fixed-width bucket quantization
+    // using the defined bucket size (1.99 × error_bound).
+    var encoded_value: u64 = 0;
+    for (uncompressed_values) |value| {
+        if (!math.isFinite(value) or @abs(value) > tester.max_test_value) return Error.UnsupportedInput;
+
+        if (error_bound == 0.0) {
+            // Quantization for the lossless case.
```

Comment:
As described in Section 7.1.2 of the paper, the source code is available in [[7]](https://github.com/Spatio-Temporal-Lab/Serf). So maybe check the source code for Serf to see how they have implemented this case. Also, it may also be worth checking how they handle floating-point arithmetic, maybe `1.99` can be changed to `2` without compressing some values incorrectly.

Reply:
Seems a few places still say `1.99` and even `1.999`, is this on purpose?